### PR TITLE
feat: refactor deno_typescript, add incremental builds, move bundling to Rust

### PR DESCRIPTION
**Pull request by kitsonk**
Ref #6686 (it does it for deno_typescript but not `deno bundle`)

Raising this to give visibility.

This is a pretty major refactor of deno_typescript and how we create bundles.  It does this:

- Provides incremental/build info support to `deno_typescript`
- Refactors and cleans up a lot of `compiler_main.js` in `deno_typescript`
- Moves the bundling logic away from TypeScript to Rust

In order to accomplish this, there is now a `main.tsbuildinfo`, `compiler.tsbuildinfo`, `main.cache`, and `compiler.cache` that are checked into source control.  When we are running an incremental compiler, only modules that have changed get emitted, so this means in order to generate the bundle, we need a "cache" of previous emitted files so that we can combine that with any change (emitted) files.  We could store them in the target output, but it would mean that everyone would have to build from source when building.  By storing the build info and the cache in source control, we should be able to speed up all builds.

The other "challenge" with bundling, is that we need an easy way to "bundle" source maps, so that is now in deno_typescript.

This pattern to bundling can like become a feature of the cli as well, which maybe means we actually make it part of core?  I still think it is a little too opinionated and needs to be a bit more flexible to be easily integrated into the CLI, but since I got it all working, I wanted to raise the PR for sharing/feedback.

### CODE REVIEW

**Comment by kitsonk** [comment]
Holy cow... that can really speed things up.  Build release on Mac for this PR was a little over 4 minutes... the build of the last commit on master was a little over 20 minutes.  🤯

**Comment by kitsonk** [comment]
Actually look at some other builds... it actually usually hovers around 6m, sometimes taking 20 minutes.  That is a bit more like what I expected it to be, a 30% savings.

**Comment by ry** [comment]
It's definitely much faster - impressive!
```
touch cli/js/web/fetch.ts; cargo build

master           2m 06s
master           1m 55s
deno_ts_bundle      54s
deno_ts_bundle      51s
```
I'm not so excited about checking in the tsbuildinfo files, nor v excited about digging in with deno_typescript. I want to reduce complexity in the internal code base, at the expense of type safety, because there are all sorts of stuff that we don't have a handle on in cli/js. That said, I don't have a very good strategy for nuking deno_typescript right now - Bartek and I tried yesterday - and immediately ran into problems. That in itself is worrying. So... I'm not sure about this patch.

**Comment by kitsonk** [comment]
Personal opinion, it takes some pressure off the build process, as part of a longer term evolution.  It does minimal changes to the API surface area of deno_typescript, so even if it doesn't last long, it essentially zero cost to accept it now.  I mean 4 extra files in the source tree to speed up builds for now while other things likely will be a higher priority seems like a small price to pay.  It also makes the code overall far more maintainable, IMO, even if it ends up getting thrown away.

Even if we don't merge it, it was a good learning experience, as it proves out a couple things:

- We can pull bundling fully out of JavaScript/TypeScript, we just need something that emits the modules, which makes transpiling in swc much more viable.
- It is easier than I thought to manipulate/bundle source maps.

It gave me enough information, that what I would want to do is potentially totally nuke the current compiler process in the cli and majorly refactor.  I think we all agree it has grown organically over the past two years and it is super hard to follow and it doesn't really align to moving more and more stuff to rust.  It would take a while.

Actually thinking out loud, I think I could create a single TypeScript compiler crate...  While the caching for example would still remain in the CLI (for now) the TypeScript compiler would just be focused on taking TypeScript and providing JavaScript output, including bundles of JavaScript.  That would abstract out a whole load of stuff, it also means it could be a clean implementation and be integrated when ready.  How does that sound?

**Comment by ry** [comment]
>  It does minimal changes to the API surface area of deno_typescript, so even if it doesn't last long, it essentially zero cost to accept it now. 

Yes - good point. @bartlomieju you ok with merging this?

**Comment by bartlomieju** [comment]
I did a small test: first I built this branch, then I added this line at the end of `cli/js/compiler.ts`: `// console.log("hello");`. I rebuilt the project and run `git status`:
```
▶ gs
On branch deno_ts_bundle
Your branch is up to date with 'kitsonk/deno_ts_bundle'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   cli/js/compiler.cache
	modified:   cli/js/compiler.tsbuildinfo
```
Then I removed added line and built project again, then run `git status` again:
```
▶ gs
On branch deno_ts_bundle
Your branch is up to date with 'kitsonk/deno_ts_bundle'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   cli/js/compiler.cache
```
It seems builds are not idempotent, probably some path gets changed in `.cache` file. 

I'm in favor of this change, but do we really need to commit `.buildinfo` and `.cache` files into repo? CI run time is not of concern here, just DX locally.

**Comment by kitsonk** [comment]
The `.cache` does not care about ordering... so what happened there is the position of the `js/compiler.ts` moved in the cache from its original position, because it was the only file emitted.  There could be extra work to preserve key ordering but that seems unnecessary to me.

While CI run time isn't a primary concern I would think it is a concern none the less.  I am not bothered either way, but the advantage of checking them into the repo also benefits local DX.  Basically when you pull in something in, you are using someone else's computer to do the transpiling once, instead of in a distributed way, so pulling in a big change to `/js/` from somewhere else would not add local build time.

Of course I am not wed to it and it is a simple change.  I just realised too that if we do keep them checked in, I need to "pretty" format the `.cache` so it is on multiple lines, so that the diffs generated are more manageable.

**Comment by kitsonk** [comment]
Useless now.  Will continue work in a stand-alone compiler crate.
